detect/decoder.go

```diff
@@ -0,0 +1,153 @@
+package detect
+
+import (
+	"bytes"
+	"encoding/base64"
+	"regexp"
+	"unicode"
+)
+
+var b64Regexp *regexp.Regexp
+var decoders = []func(string) ([]byte, error){
+	base64.StdEncoding.DecodeString,
+	base64.RawURLEncoding.DecodeString,
+}
+
+// EncodedSegment represents a portion of text that is encoded some way.
+// Decoded values can also have encoded text in them so there can be a kind of
+// tree of segments needing to be decoded. This is why the term parent is used
+// in a few places below.
+type EncodedSegment struct {
+	// relative (rel) vs absolute (abs) values:
+	//
+	// If a value is double encoded, multiple passes happen do decode it all.
+	// Relative values are the bounds of the encoded value in the current pass.
+	// Absolute values refer to the bounds of the original encoded segment so
+	// that we can keep track of locations in the original data.
+	//
+	relStart     int
+	relEnd       int
+	absStart     int
+	absEnd       int
+	decodedValue string
+}
+
+func init() {
+	b64Regexp = regexp.MustCompile(`[\w\/\-\+]{16,}={0,3}`)
```

Comment:
Just curious, why do this in an init block instead of directly assigning it?

Reply:
Ah, didn't realize I could. I'll tweak that. Good catch.

Reply:
I'll also change the other place where I did that as well.